### PR TITLE
add basic directory structure and initial documentation

### PR DESCRIPTION
**Pull request by dkochmanski**
- data frame protocol specification and stub implementation

### CODE REVIEW

**Comment by dkochmanski** [comment]
@skempf @slyrus if you have feedback on this please leave them in form of a review

**Comment by dkochmanski** [comment]
@sirherrbatka ^

**Comment by sirherrbatka** [comment]
This protocol honestly looks fine for it's intended purpose. The only thing that bothers me, and it is not because it is objectively wrong, is that only strings can be used as column/row names. I already decided to use symbols for this purpose instead. Would it be possible to extend the name type to be (or string symbol)? This would make it slightly more complicated because nil is a symbol. In fact, I am starting to consider revisiting my own decision now…

Do you plan to make some kind of spreadsheet viewer for the frames as well? I would be very interested in that!

**Comment by dkochmanski** [comment]
what would be a benefit of using symbols? I can see only eql comparisons, but if we allow also strings, then you need to use string= anyway. NIL is not a problem here.

Note, that I've dropped notion of named rows based on yours and Kabriel's suggestions.

Regarding spreadsheet viewer, crude implementation is easy:

```
(in-package clim-user)

(defun show-data-frame (clim-stream data-frame)
  (formatting-table (clim-stream)
    (formatting-row (clim-stream)
      (surrounding-output-with-border (clim-stream :background +grey+
                                                   :padding-top 0
                                                   :padding-bottom 0)
        (formatting-cell (clim-stream)
          (princ "id" clim-stream))
        (map nil (lambda (column-name)
                   (formatting-cell (clim-stream)
                     (princ column-name clim-stream)))
             (eu.turtleware.polyclot:cols data-frame))))
    (eu.turtleware.polyclot:map-data-frame-rows 
     data-frame t
     (lambda (row-index row)
       (formatting-row (clim-stream)
         (formatting-cell (clim-stream)
           (print row-index clim-stream))
         (eu.turtleware.polyclot:map-data-frame-cols
          data-frame row t
          (lambda (col-index col-name value)
            (declare (ignore col-index col-name))
            (formatting-cell (clim-stream)
              (print value clim-stream)))))))))

(define-application-frame poor-mans-spreadsheet ()
  ((dataframe :initarg :dataframe :reader dataframe))
  (:pane :application :text-margins '(:left 10 :top 5)
   :display-function (lambda (frame pane)
                       (show-data-frame pane (dataframe frame)))))

(run-frame-top-level
 (make-application-frame
  'poor-mans-spreadsheet
  :dataframe
  (eu.turtleware.polyclot:make-data-frame
   '("name" "col1" "col2" "col3")
   '("row1" value1 value2 value3)
   '("row2" value1 value2 value3))))
```

obraz

With `slim` utilities that would be even shorter.

**Comment by sirherrbatka** [comment]
As for symbols, it would take me a moment to explain exactly why but it made implementation slightly easier for me. I agree that for purpose of this protocol there is no advantage.

I will need time to digest spreadsheet example but thanks! As simplistic as it may be, It should be usefull.

File: Source/dataframe.lisp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                   for row in (rows df)
+                   for val = (funcall fun ind row)
+                   do (vector-push-extend val row)))))
+
+(defun make-data-frame (cols &rest rows)
+  (let* ((new-cols (copy-array (coerce cols 'vector) :adjustable t))
+         (new-rows (make-array (length rows) :adjustable t))
+         (col-width (length new-cols))
+         (index 0))
+    (map nil (lambda (row)
```
Comment (skempf) [inline Source/dataframe.lisp:255]: Would it be worth checking the columns here too?
`
    (assert (not (= col-width (length (remove-duplicates cols))))
            nil 
            '<col-name-not-unique>)
`

File: Source/dataframe.lisp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+               do (funcall fun index row)))))))
+
+(defgeneric map-data-frame-cols (<data-frame> %row-index% %col-slice% function)
+  (:method ((df <raw-data-frame>) row col-slice fun)
+    (assert (typep row '%row-spec%)        nil '<row-invalid-index>)
+    (assert (typep col-slice '%col-slice%) nil '<col-invalid-slice>)
+    (typecase row
+      (integer (setf row (elt (rows df) row)))
+      ;; Undocumented hack to map over col names.
+      (null (setf row (make-array (length (rows df))))))
```
Comment (skempf) [inline Source/dataframe.lisp]: Reading an uninitialized element of an array is undefined. Maybe just change to
`
(null (setf row (make-array (length (rows df)) :initial-element nil))))
`

File: Source/dataframe.lisp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                 do (funcall fun index name value)))
+          ((cons (eql t) %col-index%)
+           (loop with i2 = (norm-index (cdr col-slice))
+                 for index from 0 upto i2
+                 for name across cols
+                 for value across row
+                 do (funcall fun index name value)))
+          ((cons %col-index% (eql t))
+           (loop with i1 = (norm-index (car col-slice))
+                 for index from i1
```
Comment (skempf) [inline Source/dataframe.lisp]: should be
`                 for index from i1 below (length (cols df))
`

File: Source/dataframe.lisp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        ((eql t)
+         (loop for index from 0
+               for row across rows
+               do (funcall fun index row)))
+        ((cons (eql t) %row-index%)
+         (loop for index from 0 upto (cdr row-slice)
+               for row across rows
+               do (funcall fun index row)))
+        ((cons %row-index% (eql t))
+         (loop for index from (car row-slice)
```
Comment (skempf) [inline Source/dataframe.lisp]: should be
`
         (loop for index from (car row-slice) below (length (rows df))
`

File: Documentation/documentation.org [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+Polyclot is a tool to draw interactive charts in CLIM. Purpose of this
+document is to provide information on how to plot data and how to
+create new types of charts. Tool may be used by other CLIM
+applications as a library or as a standalone utility to render charts.
+
+* Developer guide
+** Coding conventions
+- class names are enclosed in brackets, i.e ~<data-frame>~
+- type names are enclosed in percentage characters, i.e ~%index%~
+- anaphoric macro variable names shoud start with dash, i.e ~^it~
```
Comment (skempf) [inline Documentation/documentation.org]: Do you mean caret (instead of dash), or is it supposed to be ~-it~?

File: Documentation/documentation.org [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+**** <invalid-index> (error)                                      :condition:
+**** <row-invalid-index> (<invalid-index>)                        :condition:
+**** <col-invalid-index> (<invalid-index>)                        :condition:
+**** <row-does-not-exist> (<row-invalid-index>)                   :condition:
+**** <col-does-not-exist> (<col-invalid-index>)                   :condition:
+**** <insert-error> (error)                                       :condition:
+**** <col-name-not-unique> (<insert-error>)                       :condition:
+**** <row-length-mismatch> (<insert-error>)                       :condition:
+*** Accessors                                                            :gf:
+**** dims <data-frame>
```
Comment (skempf) [inline Documentation/documentation.org:142]: Why two values and not a cons or something like it, more similar to array-dimensions?

**Review by skempf — CHANGES_REQUESTED** [review]
This all looks really good to me. Just a few minor fixes below, mixed with some comments.

For future reference, why did you switch from MIT to BSD (I don't care either way)?

Also, I was wondering if we need to put bound checking on slices, or is it better to let the error get tripped by elt or whatever access function is being used.

File: Documentation/documentation.org (hunk 5 above)
Comment (dkochmanski) [reply Documentation/documentation.org]: I do mean carret, thanks (fixed)

File: Documentation/documentation.org (hunk 6 above)
Comment (dkochmanski) [reply Documentation/documentation.org:142]: I didn't think about it really. It avoids consing I suppose. Should I change it?

File: Source/dataframe.lisp (hunk 4 above)
Comment (dkochmanski) [reply Source/dataframe.lisp]: good catch, fixed

File: Source/dataframe.lisp (hunk 3 above)
Comment (dkochmanski) [reply Source/dataframe.lisp]: fixed

File: Source/dataframe.lisp (hunk 2 above)
Comment (dkochmanski) [reply Source/dataframe.lisp]: fixed

File: Source/dataframe.lisp (hunk 1 above)
Comment (dkochmanski) [reply Source/dataframe.lisp:255]: I think you mean

```
(assert (= col-width (length (remove-duplicates cols :test #'string=)))
            nil
            '<col-name-not-unique>)
```

yes, good idea. fixed.

File: Source/dataframe.lisp (hunk 2 above)
Comment (dkochmanski) [reply Source/dataframe.lisp]: after a thought I've changed it to simply use cols (the arrity always matches and it makes sense to pass the column name as a value).

**Comment by dkochmanski** [comment]
I've added another commit with a tutorial

**Comment by skempf** [comment]
Are you force pushing updates? Does that mean when I check out, I need to rebase? I'm not making changes on my end, just reviewing, but every time I checkout, it has merge conflicts.

File: Documentation/documentation.org (hunk 6 above)
Comment (skempf) [reply Documentation/documentation.org:142]: That would be my suggestion. Unless I am thinking about it incorrectly, with the second dimension being the second value, you would always need to do a multiple-value-bind. As a cons, it is more like the `#'array-dimensions`.

**Comment by dkochmanski** [comment]
I am. please try

```
git fetch origin
git reset --hard origin/branch-name
```

Note that I'm rebasing the pull request (not the master branch), to have a clean merge. I wouldn't change `master` history.

**Comment by dkochmanski** [comment]
I've rebased once more. I've added also stub tests.

File: Documentation/documentation.org (hunk 6 above)
Comment (dkochmanski) [reply Documentation/documentation.org:142]: getting list is simple: (multiple-value-list (dims df)), if you want to access a specific value, you may call (nth-value 1 (dims df))

Comment (skempf) [reply Documentation/documentation.org:142]: Quite right. Using multiple values always requires a wrapper around the function that understands multiple values, which is why I generally avoid it, but this is perhaps a wrongly based bias. Some (pjb) have suggested that "always use multiple values!". So it is!

File: Documentation/doc-dataframe.lisp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+;;; This example file follows the data frame tutorial
+;;; snippets. SHOW-DATA-FRAME is a helper utility for data frames to
+;;; visualise them. If you load this file it will open a few windows,
+;;; it is advised to evaluate top level forms step by step when
+;;; reading the tutorial.
+
+(defun show-data-frame (data-frame
+                        &aux (clim-stream (clim:open-window-stream)))
+  (setf (clime:stream-text-margins clim-stream) '(:left 10 :top 5))
```
Comment (skempf) [inline Documentation/doc-dataframe.lisp:13]: I'm using the version of McCLIM in quicklisp. In this version, `'(:left 10 :top 5)` is not of type margin-spec. I have to edit it to `'(:left (:absolute 10) :top (:absolute 5))`.

File: Documentation/doc-dataframe.org [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+               (declare (ignore row-index))
+               (let ((address (ref *data-frame* row "address")))
+                 (switch (address :test (lambda (elt pat)
+                                          (member elt pat :test #'string=)))
+                   ('("Franz, Inc." "Symbolics, Inc." "Lucid, Inc.") "USA")
+                   ('("TurtleWare" "Just, Inc.")                     "Poland")
+                   (otherwise                                        "Unknown")))))
+#+END_SRC
+
+Notice that we've used a function ~ref~. We could have used ~elt~
```
Comment (skempf) [inline Documentation/doc-dataframe.org]: But shouldn't we prefer `ref`, because using `elt` exposes the implementation? (this applies here and in the code snippet here and in doc-dataframe.lisp).

File: Documentation/doc-dataframe.org [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+Mapping is prefered over selection because it doesn't cons new
+sequences. Slice specifiers are the same for both operations. It is
+possible to map over whole data frame, over rows and over a single row
+column-by-column. Example for printing all names in a data frame:
+
+#+BEGIN_SRC common-lisp
+(map-data-frame *data-frame* t #("name")
+  (lambda (rind row cind cname value)
+    (declare (ignore rind row cind cname))
+    (print name)))
```
Comment (skempf) [inline Documentation/doc-dataframe.org]: `(print value)))`

**Review by skempf — APPROVED** [review]

**Comment by skempf** [comment]
I'm not sure how long you want to keep it in review. I've read through some of the documentation several times now, and I forget what has changed and what has not, because I can't diff a previous commit. There are a few grammar errors, but the github gui makes it too difficult to fix that. I could send you a modified org file or something.

File: Documentation/doc-dataframe.lisp (hunk 7 above)
Comment (dkochmanski) [reply Documentation/doc-dataframe.lisp:13]: next version will allow atomic values. either way it is not a file which is loaded with the system.

File: Documentation/doc-dataframe.org (hunk 8 above)
Comment (dkochmanski) [reply Documentation/doc-dataframe.org]: I've removed the paragraph.

File: Documentation/doc-dataframe.org (hunk 9 above)
Comment (dkochmanski) [reply Documentation/doc-dataframe.org]: fixed, thank you.

**Comment by dkochmanski** [comment]
sending a diff will work fine for me. regarding review, I want to have at least basic tests for all functions mentioned in the spec.

**Comment by skempf** [comment]
That leaves the map-X functions and add-X! functions. I might have some time this evening to work on it, if you don't have it done by then!

**Comment by dkochmanski** [comment]
Sounds good to me, I'll push sel and ref in two hours

**Comment by dkochmanski** [comment]
I've pushed (without rebasing to not introduce unnecessary merge conflicts to you -- I'll squash in the end) tests for ref and sel. They show some problems for invalid data.

**Comment by skempf** [comment]
gen-valid-row produces a row of single type data using a random index into a case table. When generating multiple rows, this function is called multiple times and produces rows of varying types (from row to row, but single type within the row). I had an issue with this, because when adding a new column with add-cols!, there are some instances where this throws an error. One example is if a given row is a bit vector, and then the new column data is a string, it fails trying to vector-push-extend the string into the bit vector.

Really, data-frames should be column type consistent, such that a given column is of the same data type.

Or it could be that when you said "row-based" originally, you meant to have a data-frame that is 90 deg rotated from what one might normally define (or at least that I'm used to). I have some ideas, but this question has to be answered first. Not sure if it is easier to discuss this here, or over irc.

Here is an example of a add-cols! test that shows the problem
```
(def-test add-cols! ()
  (for-all ((width  (gen-integer :min 2 :max +max-cols+))
            (height (gen-integer :min 1 :max +max-rows+))
            (new-width (gen-integer :min 1 :max +max-cols+)))
    (let* ((df (funcall (gen-data-frame height width)))
           (new-col-spec (funcall (gen-valid-cols new-width :prefix "new")))
           (new-col-data (funcall (gen-valid-rows height 1)))
           (name-fun-pairs (loop
                             :for ind :from 0 :below new-width
                             :append (list (elt new-col-spec ind)
                                           (lambda (row-index row)
                                             (declare (ignore row))
                                             (elt new-col-data row-index))))))
      (apply #'add-cols! df name-fun-pairs)
      (is (= (nth-value 1 (dims df)) (+ width new-width))))))
```

A dirty fix is to add `:element-type t` to the row copy-array in `make-data-frame`. But I really started thinking about the end use. As I currently understand the specification, I'm going to want to plot columns (not rows). It seems that we need better access to views of columns, and it would likely be better to arrange the data as columnar arrays rather than row-arrays.

If you want it to be row-based, with homogeneous rows, then the rows need to be named and the columns would represent data points. I'm UTC-5, so I might catch you early morning, or can try to chat your tomorrow afternoon.

**Comment by skempf** [comment]
I tried to un-approve my approval. It looks like I requested myself to review it again. Sorry if I messed it up.

**Comment by dkochmanski** [comment]
let's consider i.e a scatter plot, with rows as points I see it as:

```
|  x |  y |
|----+----|
|  1 |  2 |
|  3 |  1 |
|  4 |  5 |
| 12 | 42 |
```

plotting it would be something in this spirit:

```
(map-over-data-frame-rows df t
  (lambda (index row)
    (draw-point (ref df row "x") (ref df row "y"))))
```

Now we could plot a time series, where Y is the value and X is ignored:

```
(let ((x 0))
  (map-over-data-frame-cols df t "y"
    (lambda  (cind cname value)
      (draw-point x value)
      (incf x))))
```

how would it look like with the alternative representation? I think that a few examples of a use case would help me to understand the use case better. We can modify the specification to allow frames with different arrangement; also it is not that we are not allowed to change the specification when we encounter problems with it later down the road.

I gather you mean something like this?
```
| name | p1 | p2 | p3 | p4 | ... | pn |
|------+----+----+----+----+-----+----|
| x    |  1 |  2 |  3 |  4 | ... | n  |
| y    |  1 |  4 |  9 | 16 | ... | n² |
```

It is worth noting, that specialized data frames are not prohibited by the spec, it is only that they are left unspecified. i.e we could create a class `<readonly-specialized-data-frame>` which errors on add-rows! and add-cols! suggesting that operations should be performed on the data frame copy (and copy-data-frame would return an "ordinary" `<raw-data-frame>`) - then all rows (or cols) could be specialized non-adjustable vectors.

**Comment by dkochmanski** [comment]
these pushed merges with "solved conflicts" are quite troublesome. please consider rebasing your changes before push next time. I'll try to fix the history (please reset --hard to origin then).

**Comment by skempf** [comment]
In the function `gen-valid-row-slice`, case 4 sets up an array of indices to be taken as the slice. The length is `(random (dims data-frame))`, which sometimes results in a slice that is a zero length array `#()`. Is this really a valid case? I couldn't think of a use, but maybe?

`map-data-frame-rows` will accept any invalid function if the row-slice is an empty array. For now, I changed the creation of a valid row slice to use `(max 1 (random (dims data-frame)))` so my test would work as-is. 

I have some changes for you to look at, but I'll wait until tomorrow to push them.

**Comment by skempf** [comment]
Regarding your use case example above. I think you are correct. Your first table and example plot functions are what I had in mind. I was reading the implementation backwards, but I have straightened up now!

**Comment by dkochmanski** [comment]
0-length arrays are fine (imagine an "empty" data frame to which we add columns later). Just like you can map over empty sequence (function is not called at all).

**Comment by skempf** [comment]
Big caveat: this is the first time I have used fiveam.

I setup a function to generate invalid functions to pass to `add-cols!`. This could also be used for `map-data-frame-rows`, because that function also takes a function of two arguments (row-index row).

When I construct the `map-data-frame-rows.invalid` test, I supply it with a combination of valid and invalid arguments. In particular, the combination of supplying a valid row-splice that happens to be an empty array and an invalid function, does not signal an error. This is because the function is never run, as you point out above.

I don't think it would be time well spent to put the required logic into `map-data-frame-rows` to test the function, even if the row-splice is empty. My inclination is to prevent that particular corner case when running the test for `map-data-frame-rows`.

**Comment by dkochmanski** [comment]
(for-all …) bindings have an optional third clause called guard - when guard form yields nil body is not executed. you could do it like this:

```
(for-all ((valid-rows (gen-valid-row))
          (invalid-cols (gen-invalid-col) (not (emptyp valid-rows))))
  (is (whatever)))
```

then you do not have to create a separate generator nor it has to be altered to return smaller subset of valid rows.

**Comment by skempf** [comment]
Thanks; that works great. Just have one more to write. Are you OK if I push the changes? There aren't any changes to origin, so I should be OK.

**Comment by dkochmanski** [comment]
did you try git fetch + git rebase ? either way it should be OK

**Comment by skempf** [comment]
There might be a small leak in that the guard blanking is combined with other tests in `map-data-frame[-X].invalid`, because I think none of the tests for that set of inputs is run if the guard is nil (i.e. empty row or col slice). If the number of tests is high enough, it still should catch all possibilities.

File: Tests/test-dataframe.lisp [hunk 10]
```diff
@@ -271,6 +271,71 @@
                                  (funcall gen-index*))
                            (return array))))))))
 
+
+
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: please remove unnecessary empty lines (one empty line between toplevel forms is plenty, if you feel that things do not belong to the same page insert a break page character -- in emacs C-q C-l)

File: Tests/test-dataframe.lisp [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+    (signals <invalid-slice> (map-data-frame-cols df row-spec* col-slice* function))
+    (signals <invalid-slice> (map-data-frame-cols df row-spec* col-slice* function*))))
+
+(def-test add-rows! ()
+  (for-all ((width  (gen-integer :min 1 :max +max-cols+))
+            (height (gen-integer :min 1 :max +max-rows+))
+            (new-height (gen-integer :min 1 :max +max-rows+)))
+    (let ((df (funcall (gen-data-frame height width)))
+          (new-row-spec (funcall (gen-valid-rows new-height width))))
+      (apply #'add-rows! df new-row-spec)
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: add finishes:

    (finishes (apply #'add-rows! df new-row-spec))

File: Tests/test-dataframe.lisp [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+          (new-row-spec (funcall (gen-invalid-rows new-height width))))
+      (signals <insert-error> (apply #'add-rows! df new-row-spec)))))
+
+(def-test add-cols! ()
+  (for-all ((width  (gen-integer :min 1 :max +max-cols+))
+            (height (gen-integer :min 1 :max +max-rows+))
+            (new-width (gen-integer :min 1 :max +max-cols+)))
+    (let* ((df (funcall (gen-data-frame height width)))
+           (name-fun-pairs (funcall (gen-valid-name-fun-pairs new-width))))
+      (apply #'add-cols! df name-fun-pairs)
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: ditto

File: Tests/test-dataframe.lisp [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-invalid-row-slice df))))))
+            (function (gen-valid-function 'row))
+            (function* (gen-invalid-function 'row) 
+                       (if (and (arrayp row-slice)
+                                (zerop (array-total-size row-slice)))
+                           nil
+                           t)))
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: replace if with just a condition:
`(not (and (arrayp …) (zerop …))`

File: Tests/test-dataframe.lisp [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+                       (funcall (gen-invalid-row-slice df))
+                       (funcall (gen-invalid-col-slice df))))))
+            (function (gen-valid-function 'map))
+            (function* (gen-invalid-function 'map)
+                       (if (or (and (arrayp row-slice)
+                                    (zerop (array-total-size row-slice)))
+                               (and (arrayp col-slice)
+                                    (zerop (array-total-size col-slice))))
+                           nil
+                           t)))
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: replace if with just a test

`(not (or (and ) (and ))`

File: Tests/test-dataframe.lisp [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+                      (cons (elt row 0) (elt row 1))
+                      42)))
+           (col (lambda (col-index col-name value)
+                  (declare (ignore col-index))
+                  (cons col-name value)))
+           (map (lambda (row-index row col-index col-name value)
+                  (declare (ignore row-index col-name))
+                  (eq (elt row col-index) value))))))))
+
+(defun gen-invalid-function (who)
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: add a comment explaining what is `who` variable and what are its possible values.

File: Tests/test-dataframe.lisp [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+  (lambda ()
+    (case (random 3)
+      ;; not a function
+      (0 42)
+      ;; wrong parameter definition
+      (1 (lambda (only-one-arg)
+           only-one-arg))
+      ;; invalid use of arguments
+      (2 (case who
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: I suppose that this should be `ecase`?

File: Tests/test-dataframe.lisp [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+             (loop for i from 0 below length
+                   do (setf (elt array i)
+                            (funcall gen-index))
+                   finally (setf (elt array wrong)
+                                 (funcall gen-index*))
+                           (return array))))))))
+
+
+
+(defun gen-valid-function (who)
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: explain who argument in comment

File: Tests/test-dataframe.lisp [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+  (lambda ()
+    (case (random 2)
+      ;; constant
+      (0 (constantly 42))
+      ;; combination of inputs
+      (1 (case who
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: ecase?

**Review by dkochmanski — COMMENTED** [review]
some minor change requests

File: Tests/test-dataframe.lisp [hunk 19]
```diff
@@ -0,0 +1,554 @@
...
+                                 (funcall gen-index*))
+                           (return array))))))))
+
+
+
+(defun gen-valid-function (who)
+  (lambda ()
+    (case (random 2)
+      ;; constant
+      (0 (constantly 42))
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: is constantly used here because it accepts any number of arguments?

File: Tests/test-dataframe.lisp [hunk 20]
```diff
@@ -0,0 +1,554 @@
...
+      (1 (case who
+           (row (lambda (row-index row)
+                  (declare (ignore row-index))
+                  (if (> (length row) 2)
+                      (cons (elt row 0) (elt row 1))
+                      42)))
+           (col (lambda (col-index col-name value)
+                  (declare (ignore col-index))
+                  (cons col-name value)))
+           (map (lambda (row-index row col-index col-name value)
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: is row/col/map always used for mapping? I think that in that case they should be named "map-row", "map-col" and "map-all" -- IMO it will be easier to grok tests for new readers.

File: Tests/test-dataframe.lisp [hunk 21]
```diff
@@ -0,0 +1,554 @@
...
+           only-one-arg))
+      ;; invalid use of arguments
+      (2 (case who
+           (row (lambda (row-index row) ; wrong index
+                  (declare (ignore row-index))
+                  (elt row (1+ (length row))))) ; 1+ for empty row
+           (col (lambda (col-index col-name value) ; very bad function
+                  (declare (ignore col-name value))
+                  (elt col-index (1+ (length col-index)))))
+           (map (lambda (row-index row col-index col-name value)
```
Comment (dkochmanski) [inline Tests/test-dataframe.lisp]: ditto (re `who` possible values -- map-row etc)

File: Tests/test-dataframe.lisp (hunk 20 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: The who=row is also used for add-cols! test (indirectly through gen-valiid-name-fun-pairs). I originally had a separate gen-(in)valid-function-X for each function (a total of 8), but the code was so similar, it looked ugly. I will work on the naming and some comments. Do you think it would actually be better with separate functions? Maybe some kind of macro for the common code and then separate to deal with each function (e.g. add-cols!, map-df-X)?

Comment (dkochmanski) [reply Tests/test-dataframe.lisp]: I think that these two functions are perfectly fine (having 8 of them would be clearly worse), just add a comment which describes what's going on and rename valid cases.

File: Tests/test-dataframe.lisp (hunk 19 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Yes. But after thinking about it, this is superfluous, because it doesn't test anything different than the following lambda function. I'm going to remove it.

File: Tests/test-dataframe.lisp (hunk 10 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Done.

File: Tests/test-dataframe.lisp (hunk 11 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Done.

File: Tests/test-dataframe.lisp (hunk 13 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Good style; thanks.

File: Tests/test-dataframe.lisp (hunk 12 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Done.

File: Tests/test-dataframe.lisp (hunk 14 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Done.

File: Tests/test-dataframe.lisp (hunk 16 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Yes.

File: Tests/test-dataframe.lisp (hunk 18 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: Yes.

File: Tests/test-dataframe.lisp (hunk 17 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: `who` is perhaps a poor word choice. I was thinking something more like this:
```
(defun gen-valid-function (which-function)
  ;; which-function is set to a symbol matching the function name we are
  ;; testing
  (lambda ()
    ;; combination of inputs
    (ecase which-function
      ((map-data-frame-rows add-cols!)
       (lambda (row-index row)
         (declare (ignore row-index))
         (if (> (length row) 2)
             (cons (elt row 0) (elt row 1))
             42)))
      (map-data-frame-cols
       (lambda (col-index col-name value)
         (declare (ignore col-index))
         (cons col-name value)))
      (map-data-frame
       (lambda (row-index row col-index col-name value)
         (declare (ignore row-index col-name))
         (eq (elt row col-index) value))))))
```

File: Tests/test-dataframe.lisp (hunk 20 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: See new commit. Changed to use name of function being tested.

File: Tests/test-dataframe.lisp (hunk 15 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: changed to `which-function` and added comment 
```
  ;; which-function is set to a symbol matching the function name we are
  ;; testing
```

File: Tests/test-dataframe.lisp (hunk 21 above)
Comment (skempf) [reply Tests/test-dataframe.lisp]: ditto

**Comment by dkochmanski** [comment]
Thank you. I've merged the branch.
